src/H5Odtype.c

```diff
@@ -123,20 +123,38 @@ const H5O_msg_class_t H5O_MSG_DTYPE[1] = {{
  *-------------------------------------------------------------------------
  */
 static htri_t
-H5O_dtype_decode_helper(H5F_t *f, unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t *dt)
+H5O_dtype_decode_helper(H5F_t *f, unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t *dt, hbool_t skip,
+                        const uint8_t *p_end)
 {
     unsigned flags, version;
-    unsigned i;
-    size_t   z;
     htri_t   ret_value = FALSE; /* Return value */
 
     FUNC_ENTER_NOAPI_NOINIT
 
     /* check args */
     HDassert(pp && *pp);
-    HDassert(dt && dt->shared);
+
+    if (!dt || !dt->shared) {
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "attempted to decode bad datatype");
+    }
```

Comment:
This should remain consistent across branches and be an assert rather than an error check

Reply:
Having this (and the matching check in [H5T_close_real](https://github.com/HDFGroup/hdf5/blob/ec4c67733fa55771edc8102a13adc2ab0b7522ad/src/H5T.c#LL3914C20-L3914C21)) as an assert raises a couple failures during testing. `H5TEST-dtypes` and `H5TEST-version_bounds_1_10` fail when accessing objects with compound datatypes that have no fields, since they expect to get an error back from the library but execution is interrupted by the assert. I think it's expecting the error to be caught earlier - I'll  look into it.